### PR TITLE
Sound fading subsystem, ambience fading.

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Adds in the ability to have sounds fades.
Makes sounds fade in and out for the ambient buzz.

## Why It's Good For The Game

It sounds so good and so smooth.

## Changelog
:cl:
add: Ambient fading
add: Sound effect subsystem
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/ambience.dm [hunk 1]
```diff
@@ -48,13 +48,17 @@ SUBSYSTEM_DEF(ambience)
 
 	if (A.ambient_buzz && (M.client.prefs.toggles & SOUND_SHIP_AMBIENCE) && M.can_hear_ambience())
 		if (!M.client.ambient_buzz_playing || (A.ambient_buzz != M.client.ambient_buzz_playing))
-			SEND_SOUND(M, sound(A.ambient_buzz, repeat = 1, wait = 0, volume = AMBIENT_BUZZ_VOLUME, channel = CHANNEL_AMBIENT_BUZZ))
+			var/sound/sound_in = sound(A.ambient_buzz, repeat = TRUE, wait = FALSE, volume = AMBIENT_BUZZ_VOLUME, channel = CHANNEL_AMBIENT_BUZZ)
+			sound_fade(sound_in, 0, AMBIENT_BUZZ_VOLUME, 20, M.client)
 			M.client.ambient_buzz_playing = A.ambient_buzz // It's done this way so I can tell when the user switches to an area that has a different buzz effect, so we can seamlessly swap over to that one
+			M.client.ambient_buzz_sound = sound_in
```
Comment (qwertyquerty) [inline code/controllers/subsystem/ambience.dm]: Should be A.ambient_buzz

**Review by qwertyquerty — CHANGES_REQUESTED** [review]

**Comment by qwertyquerty** [comment]
Fade should be a subtype of sound effect, so we can have more than one sound effect. The sound effect datum should handle all changes to the sound within itself

**Comment by PowerfulBacon** [comment]
> Fade should be a subtype of sound effect, so we can have more than one sound effect. The sound effect datum should handle all changes to the sound within itself

Alright, could make it useful for something like sounds that have their position updated as you move too

**Comment by PowerfulBacon** [comment]
@qwertyquerty updated to your specifications, pushing in a bit

File: code/controllers/subsystem/sound.dm [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	sound = S
+	listeners = _listeners
+	start_tick = world.time
+	end_tick = world.time + time
+	effect_id = generate_id()
+
+/datum/sound_effect/proc/generate_id()
+	var/id = "[name][sound.file]"
+	for(var/A in listeners)
+		id = "[id][A]"
```
Comment (qwertyquerty) [inline code/controllers/subsystem/sound.dm]: Should probably be REF(A)

File: code/controllers/subsystem/sound.dm [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		if(world.time > sound_effect.end_tick)
+			sound_effect.end_effect()
+			acting_effects -= sound_effect.effect_id
+
+		if (MC_TICK_CHECK)
+			return
+
+// ===== Sound effect procs =====
+
+/proc/sound_fade(sound/S, start_volume = 100, end_volume = 0, time = 10, var/listeners)
```
Comment (qwertyquerty) [inline code/controllers/subsystem/sound.dm:39]: I don't like this. This should all be done in the New() of /sound_effect and the New() of /sound_effect/fade

Possibly add a proc to sound effects to begin them

**Comment by ike709** [comment]
Address qwerty's comments.

**Comment by ike709** [comment]
@qwertyquerty Re-review

**Comment by ike709** [comment]
@qwertyquerty reeeeeee

**Comment by PowerfulBacon** [comment]
@qwertyquerty

**Comment by PowerfulBacon** [comment]
@qwertyquerty

**Comment by PowerfulBacon** [comment]
@qwertyquerty

**Comment by PowerfulBacon** [comment]
@qwertyquerty

**Review by qwertyquerty — APPROVED** [review]

**Pull request merged**